File: app/src/main/java/com/tianma/tweaks/miui/xp/hook/systemui/statusbar/v20201109/MiuiQSHeaderViewHook20201109.java
package com.tianma.tweaks.miui.xp.hook.systemui.statusbar.v20201109;

import android.content.Context;
import android.content.pm.PackageManager;
import android.view.Gravity;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.RelativeLayout;
import android.widget.TextView;

import com.tianma.tweaks.miui.BuildConfig;
import com.tianma.tweaks.miui.data.sp.XPrefContainer;
import com.tianma.tweaks.miui.utils.ResolutionUtils;
import com.tianma.tweaks.miui.utils.XLog;
import com.tianma.tweaks.miui.xp.hook.BaseSubHook;
import com.tianma.tweaks.miui.xp.hook.systemui.weather.WeatherMonitor;
import com.tianma.tweaks.miui.xp.hook.systemui.weather.WeatherObserver;
import com.tianma.tweaks.miui.xp.utils.appinfo.AppInfo;
import com.tianma.tweaks.miui.xp.wrapper.MethodHookWrapper;
import com.tianma.tweaks.miui.xp.wrapper.XposedWrapper;

import de.robv.android.xposed.XSharedPreferences;

/**
 * 下拉状态栏头部View Hook（下拉状态栏显示天气等)，适用版本 MiuiSystemUI(versionCode>=202011090)
 */
public class MiuiQSHeaderViewHook20201109 extends BaseSubHook implements WeatherObserver {

    private static final String CLASS_HEADER_VIEW = "com.android.systemui.qs.MiuiQSHeaderView";

    private Class<?> mHeaderViewClass;

    private Context mModContext;
    private Context mAppContext;

    private TextView mWeatherInfoTv;

    private final boolean mWeatherEnabled;
    private int mWeatherTextColor;
    private float mWeatherTextSize;

    public MiuiQSHeaderViewHook20201109(ClassLoader classLoader, XSharedPreferences xsp, AppInfo appInfo) {
        super(classLoader, xsp, appInfo);

        // mWeatherEnabled = XSPUtils.isDropdownStatusBarWeatherEnabled(xsp);
        mWeatherEnabled = XPrefContainer.isDropdownStatusBarWeatherEnabled();
        if (mWeatherEnabled) {
            // mWeatherTextColor = XSPUtils.getDropdownStatusBarWeatherTextColor(xsp);
            mWeatherTextColor = XPrefContainer.getDropdownStatusBarWeatherTextColor();
            // mWeatherTextSize = XSPUtils.getDropdownStatusBarWeatherTextSize(xsp);
            mWeatherTextSize = XPrefContainer.getDropdownStatusBarWeatherTextSize();
        }
    }

    @Override
    public void startHook() {
        XLog.d("Hooking MiuiQSHeaderView...");
        if (mWeatherEnabled) {
            mHeaderViewClass = XposedWrapper.findClass(CLASS_HEADER_VIEW, mClassLoader);
            if (mHeaderViewClass != null) {
                hookConstructor();
                hookOnFinishInflate();
                hookOnAttachedToWindow();
                hookOnDetachedFromWindow();
            }
        }
    }

    // HeaderView#constructor()
    private void hookConstructor() {
        XposedWrapper.hookAllConstructors(mHeaderViewClass,
                new MethodHookWrapper() {
                    @Override
                    protected void after(MethodHookParam param) {
                        if (mAppContext == null) {
                            View headerView = (View) param.thisObject;
                            mModContext = headerView.getContext().getApplicationContext();
                            try {
                                mAppContext = mModContext
                                        .createPackageContext(BuildConfig.APPLICATION_ID, Context.CONTEXT_IGNORE_SECURITY);
                            } catch (PackageManager.NameNotFoundException e) {
                                // ignore
                            }
                        }
                    }
                });
    }

    // HeaderView#onFinishInflate()
    private void hookOnFinishInflate() {
        XposedWrapper.findAndHookMethod(mHeaderViewClass,
                "onFinishInflate",
                new MethodHookWrapper() {
                    @Override
                    protected void after(MethodHookParam param) {
                        // MiuiQSHeaderView extends RelativeLayout
                        ViewGroup miuiQSHeaderView = (ViewGroup) param.thisObject;

                        // 右上角齿轮快捷按钮
                        ImageView shortcutView = (ImageView) miuiQSHeaderView.getChildAt(miuiQSHeaderView.getChildCount() - 1);

                        mWeatherInfoTv = new TextView(miuiQSHeaderView.getContext());
                        RelativeLayout.LayoutParams weatherInfoLp = new RelativeLayout.LayoutParams(
                                RelativeLayout.LayoutParams.WRAP_CONTENT,
                                RelativeLayout.LayoutParams.WRAP_CONTENT
                        );
                        weatherInfoLp.addRule(RelativeLayout.START_OF, shortcutView.getId());
                        weatherInfoLp.addRule(RelativeLayout.ALIGN_TOP, shortcutView.getId());
                        weatherInfoLp.addRule(RelativeLayout.ALIGN_BOTTOM, shortcutView.getId());
                        weatherInfoLp.setMarginEnd((int) ResolutionUtils.dp2px(miuiQSHeaderView.getContext(), 3));
                        mWeatherInfoTv.setLayoutParams(weatherInfoLp);

                        mWeatherInfoTv.setIncludeFontPadding(false);
                        mWeatherInfoTv.setGravity(Gravity.CENTER);

                        mWeatherInfoTv.setTextColor(mWeatherTextColor);
                        mWeatherInfoTv.setTextSize(mWeatherTextSize);

                        miuiQSHeaderView.addView(mWeatherInfoTv);
                    }
                });
    }

    // HeaderView#onAttachedToWindow()
    private void hookOnAttachedToWindow() {
        XposedWrapper.findAndHookMethod(mHeaderViewClass,
                "onAttachedToWindow",
                new MethodHookWrapper() {
                    @Override
                    protected void after(MethodHookParam param) {
                        WeatherMonitor.get(mAppContext).registerObserver(MiuiQSHeaderViewHook20201109.this);
                    }
                });
    }

    // HeaderView#onDetachedFromWindow()
    private void hookOnDetachedFromWindow() {
        XposedWrapper.findAndHookMethod(mHeaderViewClass,
                "onDetachedFromWindow",
                new MethodHookWrapper() {
                    @Override
                    protected void after(MethodHookParam param) {
                        WeatherMonitor.get(mAppContext).unregisterObserver(MiuiQSHeaderViewHook20201109.this);
                    }
                });
    }

    @Override
    public void onWeatherChanged(String newWeatherInfo) {
        XLog.d("onWeatherChanged: %s", newWeatherInfo);
        if (mWeatherInfoTv != null) {
            mWeatherInfoTv.setText(newWeatherInfo);
        }
    }
}

File: app/src/main/java/com/tianma/tweaks/miui/xp/hook/systemui/statusbar/v20201109/MiuiQSHeaderViewHook20201109.kt
package com.tianma.tweaks.miui.xp.hook.systemui.statusbar.v20201109

import android.content.Context
import android.content.pm.PackageManager
import android.content.res.Configuration
import android.view.Gravity
import android.view.View
import android.view.ViewGroup
import android.widget.RelativeLayout
import android.widget.TextView
import com.tianma.tweaks.miui.BuildConfig
import com.tianma.tweaks.miui.R
import com.tianma.tweaks.miui.data.sp.XPrefContainer
import com.tianma.tweaks.miui.utils.ResolutionUtils
import com.tianma.tweaks.miui.utils.get
import com.tianma.tweaks.miui.utils.logD
import com.tianma.tweaks.miui.xp.hook.BaseSubHook
import com.tianma.tweaks.miui.xp.hook.systemui.helper.ResHelpers
import com.tianma.tweaks.miui.xp.hook.systemui.weather.WeatherMonitor
import com.tianma.tweaks.miui.xp.hook.systemui.weather.WeatherObserver
import com.tianma.tweaks.miui.xp.utils.appinfo.AppInfo
import com.tianma.tweaks.miui.xp.wrapper.MethodHookWrapper
import com.tianma.tweaks.miui.xp.wrapper.XposedWrapper

/**
 * 下拉状态栏头部View Hook（下拉状态栏显示天气等)，适用版本 MiuiSystemUI(versionCode>=202011090)
 */
class MiuiQSHeaderViewHook20201109(classLoader: ClassLoader?, appInfo: AppInfo?) :
    BaseSubHook(classLoader, appInfo), WeatherObserver {

    private var mMiuiQSHeaderViewClass: Class<*>? = null
    private var mMiuiHeaderViewClass: Class<*>? = null

    private var mModContext: Context? = null
    private var mAppContext: Context? = null

    private var mWeatherInfoTextView: TextView? = null

    private val mWeatherEnabled: Boolean = XPrefContainer.isDropdownStatusBarWeatherEnabled
    private var mWeatherTextColor = 0
    private var mWeatherTextSize = 0f

    companion object {
        // MIUI quick settings header view
        private const val CLASS_MIUI_QS_HEADER_VIEW = "com.android.systemui.qs.MiuiQSHeaderView"

        // MIUI header view
        private const val CLASS_MIUI_HEADER_VIEW = "com.android.systemui.qs.MiuiHeaderView"
    }

    init {
        if (mWeatherEnabled) {
            mWeatherTextColor = XPrefContainer.dropdownStatusBarWeatherTextColor
            mWeatherTextSize = XPrefContainer.getDropdownStatusBarWeatherTextSize()
        }
    }

    override fun startHook() {
        logD("Hooking MiuiQSHeaderView...")
        if (mWeatherEnabled) {
            mMiuiQSHeaderViewClass =
                XposedWrapper.findClass(CLASS_MIUI_QS_HEADER_VIEW, mClassLoader)
            if (mMiuiQSHeaderViewClass != null) {
                hookConstructor()
                hookOnFinishInflate()
                hookOnAttachedToWindow()
                hookOnDetachedFromWindow()
                hookUpdateLayout()
            }

            mMiuiHeaderViewClass = XposedWrapper.findClass(CLASS_MIUI_HEADER_VIEW, mClassLoader)
            if (mMiuiHeaderViewClass != null) {
                hookShowUnimportantNotifications()
                hookShowNotificationsAnim()
            }
        }
    }

    // MiuiQSHeaderView#constructor()
    private fun hookConstructor() {
        XposedWrapper.hookAllConstructors(mMiuiQSHeaderViewClass,
            object : MethodHookWrapper() {
                override fun after(param: MethodHookParam) {
                    if (mAppContext == null) {
                        val headerView = param.thisObject as View
                        mModContext = headerView.context.applicationContext
                        try {
                            mAppContext = mModContext?.createPackageContext(
                                BuildConfig.APPLICATION_ID,
                                Context.CONTEXT_IGNORE_SECURITY
                            )
                        } catch (e: PackageManager.NameNotFoundException) {
                            // ignore
                        }
                    }
                }
            })
    }

    // MiuiQSHeaderView#onFinishInflate()
    private fun hookOnFinishInflate() {
        XposedWrapper.findAndHookMethod(mMiuiQSHeaderViewClass,
            "onFinishInflate",
            object : MethodHookWrapper() {
                override fun after(param: MethodHookParam) {
                    // MiuiQSHeaderView extends RelativeLayout
                    val miuiQSHeaderView = param.thisObject as ViewGroup

                    // 右上角齿轮快捷按钮
                    val context = miuiQSHeaderView.context
                    val shortcutViewId =
                        ResHelpers.getId(context.resources, "notification_shade_shortcut")
                    val shortcutView = miuiQSHeaderView.findViewById<View>(shortcutViewId)

                    val weatherInfoLp = RelativeLayout.LayoutParams(
                        RelativeLayout.LayoutParams.WRAP_CONTENT,
                        RelativeLayout.LayoutParams.WRAP_CONTENT
                    ).also {
                        it.addRule(RelativeLayout.START_OF, shortcutView.id)
                        it.addRule(RelativeLayout.ALIGN_TOP, shortcutView.id)
                        it.addRule(RelativeLayout.ALIGN_BOTTOM, shortcutView.id)
                        it.marginEnd =
                            ResolutionUtils.dp2px(miuiQSHeaderView.context, 3f).toInt()
                    }

                    mWeatherInfoTextView = TextView(miuiQSHeaderView.context).also {
                        it.includeFontPadding = false
                        it.gravity = Gravity.CENTER
                        it.setTextColor(mWeatherTextColor)
                        it.textSize = mWeatherTextSize
                        it.layoutParams = weatherInfoLp
                        it.id = R.id.weather_info_text_view
                    }

                    miuiQSHeaderView.addView(mWeatherInfoTextView)
                }
            })
    }

    // MiuiQSHeaderView#onAttachedToWindow()
    private fun hookOnAttachedToWindow() {
        XposedWrapper.findAndHookMethod(mMiuiQSHeaderViewClass,
            "onAttachedToWindow",
            object : MethodHookWrapper() {
                override fun after(param: MethodHookParam) {
                    val context = mAppContext ?: return
                    WeatherMonitor.get(context)
                        .registerObserver(this@MiuiQSHeaderViewHook20201109)
                }
            })
    }

    // MiuiQSHeaderView#onDetachedFromWindow()
    private fun hookOnDetachedFromWindow() {
        XposedWrapper.findAndHookMethod(mMiuiQSHeaderViewClass,
            "onDetachedFromWindow",
            object : MethodHookWrapper() {
                override fun after(param: MethodHookParam) {
                    val context = mAppContext ?: return
                    WeatherMonitor.get(context)
                        .unregisterObserver(this@MiuiQSHeaderViewHook20201109)
                }
            })
    }

    // MiuiQSHeaderView#updateLayout()
    private fun hookUpdateLayout() {
        XposedWrapper.findAndHookMethod(
            mMiuiQSHeaderViewClass,
            "updateLayout",
            object : MethodHookWrapper() {
                override fun after(param: MethodHookParam) {
                    val miuiQSHeaderView = param.thisObject as View
                    logD("MiuiQSHeaderView mLastOrientation = ${miuiQSHeaderView["mLastOrientation"]}")
                    // 当前屏幕方向
                    val orientation = miuiQSHeaderView["mLastOrientation"]
                    if (Configuration.ORIENTATION_PORTRAIT == orientation) {
                        // 竖屏
                        mWeatherInfoTextView?.visibility = View.VISIBLE
                    } else {
                        // 其他情况，隐藏天气信息
                        mWeatherInfoTextView?.visibility = View.GONE
                    }
                }
            })
    }

    // MiuiHeaderView#showUnimportantNotifications()
    private fun hookShowUnimportantNotifications() {
        XposedWrapper.findAndHookMethod(
            mMiuiHeaderViewClass,
            "showUnimportantNotifications",
            object : MethodHookWrapper() {
                override fun after(param: MethodHookParam) {
                    val headerView = param.thisObject as View
                    val weatherInfoTextView =
                        headerView.findViewById<View>(R.id.weather_info_text_view) ?: return

                    weatherInfoTextView.visibility = View.GONE
                }
            })
    }

    private fun hookShowNotificationsAnim() {
        XposedWrapper.findAndHookMethod(
            mMiuiHeaderViewClass,
            "showNotificationsAnim",
            object : MethodHookWrapper() {
                override fun after(param: MethodHookParam) {
                    val headerView = param.thisObject as View
                    val weatherInfoTextView =
                        headerView.findViewById<View>(R.id.weather_info_text_view) ?: return
                    // 显示天气信息
                    headerView.postDelayed({
                        weatherInfoTextView.visibility = View.VISIBLE
                    }, 250L)
                }
            })
    }

    override fun onWeatherChanged(newWeatherInfo: String) {
        logD("onWeatherChanged: %s", newWeatherInfo)
        mWeatherInfoTextView?.text = newWeatherInfo
    }
}